homeassistant/components/saj/sensor.py

```diff
@@ -0,0 +1,198 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_INTERVAL = timedelta(seconds=5)
+MAX_INTERVAL = timedelta(minutes=5)
+
+UNIT_OF_MEASUREMENT_HOURS = "h"
+
+SAJ_UNIT_MAPPINGS = {
+    "W": POWER_WATT,
+    "kWh": ENERGY_KILO_WATT_HOUR,
+    "h": UNIT_OF_MEASUREMENT_HOURS,
+    "kg": MASS_KILOGRAMS,
+    "°C": TEMP_CELSIUS,
+    "": None,
+}
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend(
+        {vol.Required(CONF_HOST): cv.string}, extra=vol.PREVENT_EXTRA
+    )
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up SAJ sensors."""
+
+    # Init all sensors
+    sensor_def = pysaj.Sensors()
+
+    # Use all sensors by default
+    hass_sensors = []
+
+    for sensor in sensor_def:
+        hass_sensors.append(SAJsensor(sensor))
+
+    saj = pysaj.SAJ(config[CONF_HOST])
+
+    async_add_entities(hass_sensors)
+
+    async def async_saj(event):
+        """Update all the SAJ sensors."""
+        tasks = []
+
+        values = await saj.read(sensor_def)
+
+        for sensor in hass_sensors:
+            state_unknown = False
+            if not values:
+                # SAJ inverters are powered by DC via solar panels and thus are
+                # offline after the sun has set. If a sensor resets on a daily
+                # basis like "today_yield", this reset won't happen automatically.
+                # Code below checks if today > day when sensor was last updated
+                # and if so: set state to None.
+                # Sensors with live values like "temperature" or "current_power"
+                # will also be reset to None.
+                if (sensor.per_day_basis and date.today() > sensor.date_updated) or (
+                    not sensor.per_day_basis and not sensor.per_total_basis
+                ):
+                    state_unknown = True
+            task = sensor.async_update_values(unknown_state=state_unknown)
+            if task:
+                tasks.append(task)
+        if tasks:
+            await asyncio.wait(tasks)
+        return values
+
+    async_track_time_interval_backoff(hass, async_saj)
+
+
+@callback
+def async_track_time_interval_backoff(hass, action) -> CALLBACK_TYPE:
+    """Add a listener that fires repetitively and increases the interval when failed."""
+    remove = None
+    interval = MIN_INTERVAL
+    failed = 0
+
+    async def interval_listener(now):
+        """Handle elapsed interval with backoff."""
+        nonlocal failed, interval, remove
+        try:
+            failed += 1
+            if await action(now):
+                failed = 0
+                interval = MIN_INTERVAL
+            else:
+                interval = min(interval * 2, MAX_INTERVAL)
+        finally:
+            remove = async_track_point_in_utc_time(
+                hass, interval_listener, now + interval
+            )
+
+    async_track_point_in_utc_time(hass, interval_listener, dt_util.utcnow())
```

Comment:
```suggestion
    hass.async_create_task(hass, interval_listener())
```

Reply:
I will not reply on each suggested change individually but thanks for the help!
Regarding this one to use `hass.async_create_task`: I've noticed that if the connection to the inverter is not possible at hass startup that it "blocks" the startup until the timeout appears.

When I replace it with `async_call_later(hass, 0, interval_listener)` it does not block the startup of hass itself.
Is this an option, or do you have another suggestion to prevent the blocking startup?

Reply:
Or it a possibility to do something like this: `hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)` ?
Just tested this and that does not block hass startup...